### ISSUE TITLE
[ENH] Implementing new hierarchical forecasting reconciliation methods

### ISSUE THREAD

**Issue by felipeangelimvieira** [sktime/sktime#7609]
## Context

This issue serves as a discussion topic, summarizing the current state of reconciliation methods in sktime, identifying which methods are missing, and suggesting potential improvements for user experience.

### Single-Level Approaches
**Bottom-up**: Forecast bottom levels and then aggregate to obtain forecasts for the entire hierarchy.  
**Top-down**:  
   - **Share forecast**: Transform all series (except the total) into shares relative to the total. Forecast independently, then multiply the shares by the total forecast.  
   - **Forecast proportions [7]**: Forecast all series, use the forecasts to compute shares relative to the total forecast, and then multiply by the total forecast to obtain reconciled forecasts.  
**Middle-out**: Select a level. Keep the forecasts for this level unchanged and use one reconciliation method for the series above and potentially another for the series below.

### Reconciliation Approaches

For forecast reconciliation, all series are forecasted independently, and then the base forecasts are projected onto the coherent subspace. This can be viewed as a generalized least squares (GLS) problem with linear constraints. Solutions can typically be computed using a projection matrix approach or constrained GLS. The latter is more efficient, as it requires inverting an $n_a \times n_a$ matrix instead of $n \times n$. There are various methods based on different error covariance matrix assumptions:  
   - **OLS [6]**: Simple orthogonal projection, not requiring an error covariance matrix.  
   - **OLS-Struc**: Assumes the error variance of each series is proportional to the number of series below it in the hierarchy.  
   - **Optimal reconciliation (and its variants) [2]**: Uses the error covariance matrix to compute the optimal solution, assuming constant error covariance matrices across all forecast horizons.  
   - **Reconciliation with Machine Learning [4,5]**: Uses base forecasts as inputs for machine learning algorithms, which then forecast the bottom levels. Aggregation follows to obtain the final forecasts.  
   - **Non-negative reconciliation [3]**: Solves the optimal reconciliation problem (or any other) with a constraint for non-negative forecasts. Solvable using quadratic programming (QP) algorithms. There's also an alternative heuristic that sets all negative reconciled forecasts to zero.  
   - **Reconciliation with immutable series [1]**: This method can be combined with any of the above. First, identify immutable series, then choose a set of $n_b$ basis vectors (potentially non-bottom series, ensuring the corresponding slice of the summing matrix $S$ is invertible). The summing matrix can then be partitioned for all reconciliation methods while keeping the immutable series constant. This method can also be used with non-negative reconciliation (via QP) or analytically with matrix inversion.

## Current State in sktime

Currently, the following objects are related to reconciliation:

1. **Reconciler (transformer)**: `bu`, `ols`, `wls_str`, `td_fcst`
2. **ReconcilerForecaster (forecaster)**: `mint_cov`, `mint_shrink`, `ols`, `wls_var`, `wls_str`, `bu`, `td_fcst`

The transformer applies all reconciliation methods that do not require the error covariance matrix, except for the top-down share forecasting strategy. To use the other methods, forecasting and reconciliation need to be performed simultaneously.

## Possible Improvements

I would appreciate feedback on the suggestions below:

- **All reconciliation methods should also be implemented as transformations**: In many industrial applications, optimal reconciliation is applied independently of the forecasting step. For example, different teams may generate forecasts, with a separate reconciliation strategy applied afterward using all forecasts. This is also useful in judgmental forecasting scenarios, where the error covariance matrix can be estimated from historical forecasts.  
- **Implementation of non-negative reconciliation as a hyperparameter**: This is crucial for reconciling intermittent time series, where negative forecasts may make it impossible to build a valid product after reconciliation. This feature is essential, in my opinion.  
- **Middle-out reconciliation transformer**: Potentially a transformation that accepts two reconciliation strategies and a level to serve as the "middle."  
- **Top-down share forecasting**: Although conceptually straightforward, this strategy introduces subtle API design challenges. If all methodologies are implemented as transformations, this would be a preprocessing step in a forecasting pipeline where the `transform` method casts all non-total levels to shares and `inverse_transform` applies the shares to the total forecast. This contrasts with other methods, which are simple post-processing steps where the transformation applies the reconciliation directly to the base forecasts. We may need to accept this as an outlier.

### More Complex Features

The following features may be more challenging to implement and might not be achievable as simple transformations:

- **Temporal reconciliation**: These algorithms could be extended for reconciliation across different temporal aggregations (similar to the MAPA forecaster). This could involve a transformer that takes panel/hierarchical time series and creates a hierarchical structure based on time components. However, implementing this as a transformer may require a new machine type for temporal hierarchies since the time index might not fully capture different temporal structures.

- **Cross-temporal reconciliation [7,8]**: The authors in [7] show that sequential hierarchical and temporal reconciliation does not guarantee full reconciliation. In [8], an optimal cross-temporal reconciliation strategy is proposed, where cross-sections and temporal hierarchies are reconciled simultaneously. However, it is shown that in many cases, it does not outperform standard one-way reconciled forecasts [9]. This approach seems more relevant for decision-making contexts than for accuracy improvement.

### Further Comments

1. Non-negative reconciliation requires solving an optimization problem. If this constraint is non-existent, an analytic solution can be obtained. This should be invisible to the user. If a user wants an OLS reconciliation (equivalent to an identity error covariance matrix), they should be able to choose non-negativity constraints without needing to understand the internal mechanics.
2. `ols`, `wls_var`, `wls_str`, `mint_cov`, and `mint_shrink` are essentially the same methodology given that a covariance matrix is given as input. This could help in creating a modular implementation of transformations.
3. The immutable time series approach translates the reconciliation problem to one where the chosen series remain constant. This viewpoint can be used to add this feature without introducing significant complexity.
4. For reconciliation strategies that require error covariance matrices, the transformations could accept it as an initialization hyperparameter. The `ReconcilerForecaster` could compute the matrix internally and pass it to an inner transformation if one of these methods is chosen.

## Issues

I would like to know if it would be ok to open the following issues:

1. [ENH] Add `OptimalReconciler` transformation that receives error covariance matrix as input.
2. [ENH] Add non-negative optimal hierarchical reconciliation. (implemented after or with 1.)
3. [ENH] Add MiddleOut transformation for hierarchical forecasting.
4. [ENH] Add Topdown (share) transformation for hierarchical forecasting.
5. [ENH] Add ImmutableReconciler transformation to forecast hierarchies with immutable series.

What do you think? 

References
----------

[1] Zhang, Bohan, et al. "Optimal reconciliation with immutable forecasts." European Journal of Operational Research 308.2 (2023): 650-660.  
[2] Wickramasuriya, S. L., Athanasopoulos, G., and Hyndman, R. J. (2019). "Optimal forecast reconciliation for hierarchical and grouped time series through trace minimization." Journal of the American Statistical Association, 114(526), 804–819.  
[3] Wickramasuriya, S. L., Turlach, B. A., and Hyndman, R. J. (2020). "Optimal non-negative forecast reconciliation." Statistics and Computing, 30(5), 1167–1182.  
[4] Spiliotis, Evangelos, et al. "Hierarchical forecast reconciliation with machine learning." Applied Soft Computing, 112 (2021): 107756.  
[5] Burba, Davide, and Trista Chen. "A trainable reconciliation method for hierarchical time-series." arXiv preprint arXiv:2101.01329 (2021).  
[6] Hyndman, Rob J., et al. "Optimal combination forecasts for hierarchical time series." Computational statistics & data analysis 55.9 (2011): 2579-2589.  
[7] Kourentzes, Nikolaos, and George Athanasopoulos. "Cross-temporal coherent forecasts for Australian tourism." Annals of Tourism Research 75 (2019): 393-409.  
[8] Di Fonzo, Tommaso, and Daniele Girolimetto. "Cross-temporal forecast reconciliation: Optimal combination method and heuristic alternatives." International Journal of Forecasting 39.1 (2023): 39-57.  
[9] Athanasopoulos, George, et al. "Forecast reconciliation: A review." International Journal of Forecasting 40.2 (2024): 430-456.

**Comment by fkiraly** [comment]
Some comments on the points requested

> All reconciliation methods should also be implemented as transformations

I originally thought so too, but at least a naive variant of this is not possible - with naive variant meaning, we take only forecasts as inputs of the transformation. Namely, not all reconciliation algorithms can be expressed as an algorithm in forecasts purely, e.g., covariance-based reconciliation/shrinkage where the covariance is estimated on a actual data rather than on forecasts.

There are some non-naive ways of still implementing as a transformation, namely:

* using fitted model parameters or data as `__init__` arguments
* having dedicated inputs for forecasts vs past actuals in `fit` or `predict`

but each of these does not feel good, as either we would be passing fitted parameters explicitly, or data to the `__init__`, or we would assume a very special data structure which goes beyond the base assumptions on data formats.

> Implementation of non-negative reconciliation as a hyperparameter

Agree, although it is not entirely clear to me whether to model this as a hyper-parameter or a separate estimator. I do not think all reconciliation algorithms have a "non-negative" version, typically only optimization based ones do (where you can add this as a constraint).

> Middle-out reconciliation transformer

Interesting! Agree. Although I am not sure this can always be modelled as a transformer, rather than a forecaster.

> Top-down share forecasting

I need more info on this and the thoughts you had in API design, e.g., how would pipelines or composites look like, here?

> We may need to accept this as an outlier.

... or as an indication to re-check API design assumptions ...

**Comment by fkiraly** [comment]
> I would like to know if it would be ok to open the following issues:

you are a core developer, so you can open any issues you like - the role comes with an implicit experssion of trust that you know what you are doing, and anyway anyone can open issues, but you can even manage them

If your question is, do I think these issues make sense, the answer is yes - but as said, you do not need anyone´s permission to open the issues. On PR there may be voting if there is controversy, but that rarely happens.

**Comment by felipeangelimvieira** [comment]
> I originally thought so too, but at least a naive variant of this is not possible - with naive variant meaning, we take only forecasts as inputs of the transformation. Namely, not all reconciliation algorithms can be expressed as an algorithm in forecasts purely, e.g., covariance-based reconciliation/shrinkage where the covariance is estimated on a actual data rather than on forecasts.

I'm considering the possibility of covariance matrices computed previously by the user. I can see the benefits in two scenarious:
- Judgemental forecasts: companies can have entire teams creating forecasts by hand. Based on historical data, the error covariance matrix of the forecasts can be computed.
- Using statisical forecasting without the need to re-compute error covariance: even in the case of statistical methods, the user may already have an historical of errors in their database. This can be used to avoid having to do a cross-validation to obtain such matrix. In cases where the model is expensive to train and evaluate (large databases), this can save a lot of time.

Actually, `OptimalReconciler` is a bad name. It could be something like:

```python
class ProjectReconciler(BaseTransformer):
  def __init__(self, error_covariance : pd.DataFrame = None):
    # if error_covariance is None, use error_covariance = identity matrix, which is equivalent to OLS method
    ...
```

`error_covariance` could be a dataframe with multiindex in both dimensions (rows and columns) 

In the case where **covariance matrix is computed on the fly**, indeed, I believe a forecaster could do that. It could actually leverage this "ProjectReconciler" to do it internally. 

> Agree, although it is not entirely clear to me whether to model this as a hyper-parameter or a separate estimator. I do not think all reconciliation algorithms have a "non-negative" version, typically only optimization based ones do (where you can add this as a constraint).

As far as I undestand the non-negative reconciliation, the core difference is that there is no analytical solution, and a QP algorithm is required to solve it. But the problems are essentially the same, it just requires a different "backend".

> Interesting! Agree. Although I am not sure this can always be modelled as a transformer, rather than a forecaster.

Would it be possible to pass a transformer that when applying `transform`, it keeps just one level of the hierarchy. When using `inverse_transform`, it uses the forecasts of the transformed series to recover the full hierarchy? Otherwise, a forecaster would be needed. This could also be the case for single level forecasters.

**Comment by felipeangelimvieira** [comment]
> > I would like to know if it would be ok to open the following issues:
> 
> you are a core developer, so you can open any issues you like - the role comes with an implicit experssion of trust that you know what you are doing, and anyway anyone can open issues, but you can even manage them
> 
> If your question is, do I think these issues make sense, the answer is yes - but as said, you do not need anyone´s permission to open the issues. On PR there may be voting if there is controversy, but that rarely happens.

perfect

> (on a side note, GitHub has recently introduced an "open sub-issue" feature, a button should appear at the bottom of your post. Have you tried this alerady?)

curiously I can't see it

**Comment by felipeangelimvieira** [comment]
Questions:
- Should we implement reconciliation as a transformer that goes before the model in a pipeline? If yes, should we implement a transformer that wraps a reconciler and makes it behave as it is currently (making it reconcile at `transform` instead of `inverse_transform`)?
- Is `ProjectionReconciler` also a bad name?

**Comment by fkiraly** [comment]
> Should we implement reconciliation as a transformer that goes before the model in a pipeline? If yes, should we implement a transformer that wraps a reconciler and makes it behave as it is currently (making it reconcile at `transform` instead of `inverse_transform`)?

I had to think a bit about this, sorry for the late reply. 

Overall, I think your design for the reconciler - with reconciliation in `inverse_transform` - seems more natural, given the interfaces. I wish we had thought about this at the start.

The problems fall in two categories: interface-nonconformance, and downwards compatibility.

Downwars compatibility (or lack of downwards consistency) is a problem you already mention - potentially inconsistent interfaces, and old tutorials.

There is, however, another - `transform` and `inverse_transform` are designed to be non-mutating, these methods are not allowed to write anything to `self`. Can we solve that in your design?

Further, there is also the "invert" dunder `~`, which can be used to flip `transform` and `inverse_transform` - is that something that could be used?

> Is `ProjectionReconciler` also a bad name? 

I think it is ok - although many reconcilers use some projection idea, perhaps we could be more specific?

**Comment by felipeangelimvieira** [comment]
Franz, I've created a new PR #7697, with the proposed interface. I believe this would work as a general interface, able to account for many different reconciliations. 

The PR only implement the transformations, it still do not implement the forecaster. I called the transformation that implements the optimal reconciliation `FullHierarchyReconciler`. I felt that it was more precise, since it accounts for `ols`, `wls`, and `mint`  depending on the error covariance matrix that is passed, but I'm not that good with names and would appreciate a feedback on that. I was thinking about adding the following missing blocks:
- Utilities to compute the error covariance matrix.  Possibly, a cross-validation strategy could be passed. This could be then used in `ReconcilerForecaster`.
- Add an option to use "ols" or "wls" in `FullHierarchyReconciler`, since they do not depend on the forecaster


> There is, however, another - transform and inverse_transform are designed to be non-mutating, these methods are not allowed to write anything to self. Can we solve that in your design?

Yes, they are non-mutating

> Further, there is also the "invert" dunder ~, which can be used to flip transform and inverse_transform - is that something that could be used?
The inverse_transform can be applied directly on the output of a model, but if we invert the operations, the transform should be ignored.


**If we decide to go this way, how would deprecation be?**

- Could we still use the `Reconciler` class as a shortcut to access the different reconciliation strategies, given the parameter `method`?  
- Could we add a temporary hyperparameter to define if it will behave as it is today, or according to the new proposal (inverse_transform)?

**Comment by ayushhbasu** [comment]
@felipeangelimvieira @fkiraly — thanks for the detailed discussion, this clarified a lot of the design trade-offs around hierarchical reconciliation.

From reading #7609 and related PRs (particularly #7697), it seems the main challenge is that reconciliation methods differ in their requirements: some (e.g., bottom-up/OLS) can be expressed as pure transformations on forecasts, while others (e.g., MinT/WLS) depend on additional inputs like error covariance matrices.

A direction that seems consistent with the current `sktime` design is to separate the core reconciliation logic (projection step) from the interface, and expose it via:

* a Transformer (for post-processing forecasts, optionally with user-provided covariance), and
* a Forecaster wrapper (when covariance needs to be estimated from data)

I’d be interested in contributing by:

* extending existing hierarchical reconcilers to support covariance-aware methods
* implementing additional strategies (e.g., middle-out, top-down variants)
* adding non-negative reconciliation for optimization-based methods

Happy to start with a small PR (e.g., a minimal covariance-aware reconciler prototype building on #7697) to validate the approach.

Would love to hear if this direction aligns with your intended design.

**Comment by felipeangelimvieira** [comment]
Hello @ayushhbasu !
Most of the reconcilers have already been implemented in #7697.

The reconciliation with immutable series is still missing, but requires extensive knowledge of hierarchical reconciliation. If you have prior knowledge on this topic, feel free to open a pull-request, and link to this issue: #7586

I'll close this issue in favor of this other one, and since all other methods have already been implemented,

**Comment by ayushhbasu** [comment]
Hi @felipeangelimvieira https://github.com/sktime/sktime/issues/7609#issuecomment-4138296809

Thanks! I’d like to work on immutable reconciliation (#7586).

I’m planning to start with a small prototype PR:
- add an `ImmutableReconciler` extending `FullHierarchyReconciler`
- support a simple OLS case
- implement constrained reconciliation so selected series remain unchanged while preserving coherence
- include basic tests

Once that’s validated, I can extend it to covariance-based methods (e.g., MinT/WLS) and additional constraints.

Does this direction align with your expectations?

**Comment by ayushhbasu** [comment]
Hi @felipeangelimvieira, thanks for the guidance in #7609.

I’ve opened a prototype PR #9767 for immutable reconciliation (#7586) with an OLS-based `ImmutableReconciler`.
